src/driver-gskill.c

```diff
@@ -0,0 +1,1365 @@
+	report->loop_count = 0;
+	report->please_set_me_to_1_thank_you = 1; /* No prob! Happy to help :) */
+
+	for (profile_pos = 0, increment = 1, event_idx = 0;
+	     event_idx < (signed)event_num;
+	     event_idx++, profile_pos += increment, increment = 1) {
+		event = &macro->macro.events[event_idx];
+
+		switch (event->type) {
+		case RATBAG_MACRO_EVENT_WAIT:
+			delay = (struct gskill_macro_delay*)&buf[profile_pos];
+			increment = sizeof(*delay);
+
+			delay->tag = 1;
+			delay->count = event->event.timeout;
+			break;
+		case RATBAG_MACRO_EVENT_KEY_PRESSED:
+		case RATBAG_MACRO_EVENT_KEY_RELEASED:
+			buf[profile_pos] = gskill_macro_code_from_event(device,
+									event);
+			break;
+		case RATBAG_MACRO_EVENT_INVALID:
+		case RATBAG_MACRO_EVENT_NONE:
+			goto fin;
+		}
+	}
+
+fin:
+	report->macro_length = profile_pos;
+
+	return report;
+}
+
+static struct gskill_macro_report *
+gskill_read_button_macro(struct ratbag_device *device,
+			 unsigned int profile, unsigned int button)
+{
+	struct gskill_data *drv_data = ratbag_get_drv_data(device);
+	struct gskill_macro_report *report =
+		&drv_data->profile_data[profile].macros[button];
+	uint8_t checksum;
+	uint8_t macro_num = (profile * 10) + button;
+	uint8_t buf[GSKILL_REPORT_SIZE_CMD] = { 0x0c, 0xc4, 0x0b, macro_num, 0x0 };
+	int rc;
+
+	rc = ratbag_hidraw_raw_request(device, GSKILL_GENERAL_CMD, buf,
+				       sizeof(buf), HID_FEATURE_REPORT,
+				       HID_REQ_SET_REPORT);
+	if (rc < 0) {
+		log_error(device->ratbag,
+			  "Failed to retrieve macro for profile %d for button %d: %d\n",
+			  profile, button, rc);
+		return NULL;
+	}
+
+	/* Wait for the device to be ready */
+	msleep(100);
+
+	rc = ratbag_hidraw_raw_request(device, GSKILL_GET_SET_MACRO,
+				       (uint8_t*)report, sizeof(*report),
+				       HID_FEATURE_REPORT, HID_REQ_GET_REPORT);
+	if (rc < (signed)sizeof(*report)) {
+		log_error(device->ratbag,
+			  "Failed to retrieve macro for profile %d for button %d: %d\n",
+			  profile, button, rc);
+		return NULL;
+	}
+
+	checksum = gskill_calculate_checksum((uint8_t*)report, sizeof(*report));
+	if (checksum != report->checksum) {
+		log_error(device->ratbag,
+			  "Invalid checksum on macro for profile %d button %d\n",
+			  profile, button);
+		return NULL;
+	}
+
+	return report;
+}
+
+static int
+gskill_write_button_macro(struct ratbag_device *device,
+			  struct gskill_macro_report *report)
+{
+	unsigned int profile = report->macro_num / 10;
+	unsigned int button = report->macro_num % 10;
+	uint8_t buf[GSKILL_REPORT_SIZE_CMD] = { 0x0c, 0xc4, 0x0b,
+		report->macro_num, 0x1 };
+	int rc;
+
+	rc = ratbag_hidraw_raw_request(device, GSKILL_GENERAL_CMD, buf,
+				       sizeof(buf), HID_FEATURE_REPORT,
+				       HID_REQ_SET_REPORT);
+	if (rc < 0) {
+		log_error(device->ratbag,
+			  "Failed to set macro to write for profile %d button %d: %d\n",
+			  profile, button, rc);
+		return rc;
+	}
+
+	/* Wait for the device to be ready */
+	msleep(200);
```

Comment:
Yeah :(, actually one of the first things I tried to do was figure out if there was any way we could avoid the `msleep()`s and see if we could just use the command return status for figuring out when to continue. ~~So far it hasn't seemed to work and just causes the mouse to try sending us random garbage~~ Reading this comment made me try removing the `msleep()`s just to see what would happen, and it actually looks like they might not be necessary at all for profile reads? Even after writing a shell script to continuously read from the mouse this way and check for differences in the profile data we get, it seems to always work perfectly.
Regardless, I have a feeling the results might be different if we used a mouse with an older firmware version, so it might be a good idea just to leave the delays in here anyway